APITaxi/commands/import_zupc.py

```diff
@@ -1,314 +1,388 @@
+
+    for filename in glob.glob(os.path.join(zupc_dir, '*.include')):
+        fill_zupc_update_shape(filename, parent_zupc, 'include')
+
+    for filename in glob.glob(os.path.join(zupc_dir, '*.exclude')):
+        fill_zupc_update_shape(filename, parent_zupc, 'exclude')
+
+    db.session.commit()
+
+
+def fill_zupc_tmp_table_from_arretes(zupc_repo):
+    """Fill informations in ZUPC tmp table from "arrêtés" stored in --zupc_repo.
+
+    Each folder in the root directory of the ZUPC repository contains the files describing a ZUPC.
+    """
+    for zupc_dir in os.listdir(zupc_repo):
+        fullpath = os.path.join(zupc_repo, zupc_dir)
+
+        # Skip hidden directories and extra files
+        if zupc_dir.startswith('.') or os.path.isfile(fullpath):
+            continue
+
+        fill_zupc(fullpath)
+
+
+def load_zupc_tmp_table(shape_filename, zupc_repo):
+    recreate_zupc_tmp_table()
+    fill_zupc_tmp_table_from_contours(shape_filename)
+    fill_zupc_tmp_table_from_arretes(zupc_repo)
+
+
+def merge_zupc_tmp_table():
+    """zupc_temp table contains the entries to insert into table ZUPC.
+
+    - insert all entries from zupc_temp to ZUPC (so each new ZUPC gets a new id)
+    - update the new ZUPC field "parent_id" to reference the parent
+    - update table ADS to reference the new ZUPC
+    - remove "old" ZUPC entries
+    """
+    if db.session.query(ADS) \
+                 .outerjoin(ZUPC_tmp, ADS.insee == ZUPC_tmp.insee) \
+                 .filter(ZUPC_tmp.id.is_(None)) \
+                 .count() > 0:
+        raise ValueError('Some INSEE codes referenced in table ADS do not exist in table zupc_temp')
+
+    last_zupc_id = db.session.query(func.MAX(ZUPC.id)).one()[0]
+
+    logger.info('Insert data from zupc_temp to ZUPC')
+    db.session.execute('''
+        INSERT INTO "ZUPC"(departement_id, nom, insee, shape, active)
+        SELECT departement_id, nom, insee, shape, false FROM zupc_temp
+    ''')
+
+    db.session.flush()
+
+    # For each ZUPC that has a parent in table zupc_temp:
+    # - find the parent created just before in table ZUPC
+    # - find the child
+    # - update the child with the parent's id
+    query = db.session.query(ZUPC_tmp).filter(ZUPC_tmp.parent_id != ZUPC_tmp.id)
+    logger.info('Update ZUPC.parent_id of %s rows', query.count())
+    for row in query:
+        parent_zupc = db.session.query(ZUPC) \
+                                .filter(ZUPC.id > last_zupc_id) \
+                                .filter(ZUPC.insee == row.parent.insee).one()
+
+        zupc = db.session.query(ZUPC) \
+                         .filter(ZUPC.id > last_zupc_id) \
+                         .filter(ZUPC.insee == row.insee).one()
+
+        zupc.parent_id = parent_zupc.id
+        db.session.add(zupc)
+
+    # Update ADS to reference the new ZUPC
+    insee_zupc = {
+        zupc.insee: zupc.id
+        for zupc in
+        db.session.query(ZUPC).filter(ZUPC.id > last_zupc_id)
```

Comment:
Maybe you can do a select on `id` and `insee` here, this request can be big in memory with all the shapes.

Reply:
good catch, fixed